uniref-rest/src/main/java/org/uniprot/api/uniref/queue/UniRefMessageListener.java

```diff
@@ -0,0 +1,115 @@
+            DownloadJobRepository jobRepository,
+            DownloadResultWriter downloadResultWriter,
+            RabbitTemplate rabbitTemplate) {
+        super(
+                downloadConfigProperties,
+                asyncDownloadQueueConfigProperties,
+                jobRepository,
+                rabbitTemplate);
+        this.converter = converter;
+        this.service = service;
+        this.downloadConfigProperties = downloadConfigProperties;
+        this.downloadResultWriter = downloadResultWriter;
+    }
+
+    protected void processMessage(Message message, DownloadJob downloadJob) {
+        UniRefDownloadRequest request = (UniRefDownloadRequest) this.converter.fromMessage(message);
+        String jobId = downloadJob.getId();
+        MediaType contentType = UniProtMediaType.valueOf(request.getFormat());
+        Path idsFile = Paths.get(downloadConfigProperties.getIdFilesFolder(), jobId);
+        Path resultFile =
+                Paths.get(
+                        downloadConfigProperties.getResultFilesFolder(),
+                        jobId + FileType.GZIP.getExtension());
+        // run the job if it has errored out
+        if (isJobSeenBefore(downloadJob, idsFile, resultFile)) {
+            if (downloadJob.getStatus() == JobStatus.RUNNING) {
+                log.warn("The job {} is running by other thread", jobId);
+            } else {
+                log.info("The job {} is already processed", jobId);
+                updateDownloadJob(message, downloadJob, JobStatus.FINISHED);
+            }
+        } else {
+            updateDownloadJob(message, downloadJob, JobStatus.RUNNING);
+            writeResult(request, idsFile, jobId, contentType);
+            updateDownloadJob(message, downloadJob, JobStatus.FINISHED, jobId);
+        }
+    }
+
+    private void writeResult(
+            DownloadRequest request, Path idsFile, String jobId, MediaType contentType) {
+        try {
+            writeSolrResult(request, idsFile, jobId);
+            StoreRequest.StoreRequestBuilder storeRequest = StoreRequest.builder();
+            storeRequest.addLineage(false);
+            StoreRequest storeReq = storeRequest.build();
+            downloadResultWriter.writeResult(
+                    request, idsFile, jobId, contentType, storeReq, DATA_TYPE);
+            log.info("Voldemort results saved for job {}", jobId);
+        } catch (Exception ex) {
+            logMessageAndDeleteFile(ex, jobId);
+            throw new MessageListenerException(ex);
+        }
+    }
+
+    private void writeSolrResult(DownloadRequest request, Path idsFile, String jobId)
+            throws IOException {
+        Stream<String> ids = streamIds(request);
+        saveIdsInTempFile(idsFile, ids);
+        log.info("Solr ids saved for job {}", jobId);
+    }
+
+    Stream<String> streamIds(DownloadRequest request) {
+        return service.streamIds(request);
+    }
+}
```

Comment:
The following methods are duplicate in UniRefMessageListener and UniProtKBMessageListener :
1. writeResult
2. writeSolrResult (In IdMappingMessageListener also)
3. streamIds(not major)
Move them to Abstract class and pass whatever is different e.g. DownloadResultWriter, type etc

Reply:
These improvements will be included as part of AbstractMessageListener processMessage improvement.